network/simple-onchain-discovery/src/lib.rs

```diff
@@ -0,0 +1,198 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use channel::libra_channel;
+use futures::{sink::SinkExt, StreamExt};
+use libra_config::config::RoleType;
+use libra_logger::prelude::*;
+use libra_metrics::{register_histogram, DurationHistogram};
+use libra_network_address::NetworkAddress;
+use libra_types::on_chain_config::{OnChainConfigPayload, ValidatorSet};
+use network::{
+    common::NetworkPublicKeys,
+    connectivity_manager::{ConnectivityRequest, DiscoverySource},
+};
+use once_cell::sync::Lazy;
+use std::{convert::TryFrom, time::Instant};
+
+/// Histogram of idle time of spent in event processing loop
+pub static EVENT_PROCESSING_LOOP_IDLE_DURATION_S: Lazy<DurationHistogram> = Lazy::new(|| {
+    DurationHistogram::new(
+        register_histogram!(
+            "simple_onchain_discovery_event_processing_loop_idle_duration_s",
+            "Histogram of idle time of spent in event processing loop"
+        )
+        .unwrap(),
+    )
+});
+
+/// Histogram of busy time of spent in event processing loop
+pub static EVENT_PROCESSING_LOOP_BUSY_DURATION_S: Lazy<DurationHistogram> = Lazy::new(|| {
+    DurationHistogram::new(
+        register_histogram!(
+            "simple_onchain_discovery_event_processing_loop_busy_duration_s",
+            "Histogram of busy time of spent in event processing loop"
+        )
+        .unwrap(),
+    )
+});
+
+pub struct ConfigurationChangeListener {
+    conn_mgr_reqs_tx: channel::Sender<ConnectivityRequest>,
+    role: RoleType,
+}
+
+fn extract_validator_updates(validator_set: ValidatorSet) -> Vec<ConnectivityRequest> {
+    let validator_keys = validator_set.payload().to_vec();
+    // Collect the set of address updates.
+    let updates: Vec<ConnectivityRequest> = validator_keys
+        .into_iter()
+        .map(|validator| {
+            ConnectivityRequest::UpdateAddresses(
+                DiscoverySource::OnChain,
+                validator.account_address().clone(),
```

Comment:
I had the same thought.   Taking a step back, what would be great to have is a handler which receives a NetworkConnectivityConfig from any source (on chain discovery, gossip discovery, seed file, ....) and propagates those changes into the ConnectivityManager.  That is sort of what is being done right now through the ConnectivityRequest interface, but we're writing multiple of these handlers where I think it would be simpler to write one handler that converst a NetworkConnectivityConfig into the update and a collection of ::From<> implementations that the various discovery sources implement to convert their datatype into the standard NetworkConnectivityConfig which the ConnectivityManager cares about.

Reply:
https://github.com/libra/libra/pull/4097. consolidates the UpdateAddress interface.